core/src/main/java/org/commonjava/indy/core/inject/CassandraNotFoundCache.java

```diff
@@ -0,0 +1,273 @@
+import org.commonjava.indy.model.galley.KeyedLocation;
+import org.commonjava.indy.subsys.cassandra.CassandraClient;
+import org.commonjava.maven.galley.model.ConcreteResource;
+import org.commonjava.maven.galley.model.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Default;
+import javax.inject.Inject;
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.commonjava.indy.conf.DefaultIndyConfiguration.DEFAULT_NOT_FOUND_CACHE_TIMEOUT_SECONDS;
+
+@ApplicationScoped
+@Default
+public class CassandraNotFoundCache
+                extends AbstractNotFoundCache
+{
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    private static final String TIMEOUT_FORMAT = "yyyy-MM-dd HH:mm:ss z";
+
+    private PreparedStatement preparedInsert;
+
+    private PreparedStatement preparedExistQuery;
+
+    private PreparedStatement preparedDelete;
+
+    private PreparedStatement preparedDeleteByStore;
+
+    private PreparedStatement preparedCountByStore;
+
+    private PreparedStatement preparedQueryByStore;
+
+    public static String getSchemaCreateKeyspace( String keyspace )
+    {
+        return "CREATE KEYSPACE IF NOT EXISTS " + keyspace
+                        + " WITH REPLICATION = {'class':'SimpleStrategy', 'replication_factor':1};";
+    }
+
+    // @formatter:off
+    public static String getSchemaCreateTable( String keyspace )
+    {
+        return "CREATE TABLE IF NOT EXISTS " + keyspace + ".nfc ("
+                        + "storekey varchar,"
+                        + "path varchar,"
+                        + "creation timestamp,"
+                        + "expiration timestamp,"
+                        + "PRIMARY KEY (storekey, path)"
+                        + ");";
+    }
+    // @formatter:on
+
+    @Inject
+    private CassandraClient cassandraClient;
+
+    private int maxResultSetSize; // limit the max size for REST endpoint getMissing to avoid OOM
+
+    @Inject
+    protected IndyConfiguration config;
+
+    private String keyspace;
+
+    private Session session;
+
+    protected CassandraNotFoundCache()
+    {
+    }
+
+    public CassandraNotFoundCache( final IndyConfiguration config )
+    {
+        this.config = config;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        keyspace = config.getCacheKeyspace();
+        maxResultSetSize = config.getNfcMaxResultSetSize();
+
+        session = cassandraClient.getSession();
+
+        session.execute( getSchemaCreateKeyspace( keyspace ) );
+        session.execute( getSchemaCreateTable( keyspace ) );
+
+        preparedExistQuery =
+                        session.prepare( "SELECT count(*) FROM " + keyspace + ".nfc WHERE storekey=? and path=?;" );
```

Comment:
Here, can we do something like a count(1) or select 1 or similar, so we can return faster when we find a match?  The actual count appears irrelevant as long as we know there is at least one. Will that help performance at all?

Reply:
Google it and seems the count(*) is better. The * is not evaluated so it is the fastest. Ref to https://stackoverflow.com/questions/181272/what-does-select-count1-from-table-name-on-any-database-tables-mean